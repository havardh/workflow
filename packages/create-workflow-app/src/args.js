/* eslint-env node */
/* eslint-disable no-console */
function usage() {
  console.log(`Initialize a new workflow app

Usage:
  npm init workflow-app [target]
  npx create-workflow-app [target]
  yarn create workflow-app [target]

Initializes the current directory or target if given
with the create-workflow-app initializer.
`);
}

export default function parse(args) {
  if (args.length < 1) {
    console.error('Invalid number of arguments');
    console.log();
    usage();
    process.exit(1);
  }

  const [node, cmd, targetFolder] = args; // eslint-disable-line

  if (args.length === 2) {
    return {};
  } else if (args.length === 3) {
    return { targetFolder };
  } else {
    console.error("Invalid arguments: '" + args.join(' ') + "'");
    console.log();
    usage();
    process.exit(1);
  }
}
