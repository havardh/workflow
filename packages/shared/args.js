export default function args(args) {
  const parsedArgs = {
    named: {},
    positional: [],
  };

  const argsLength = args.length;
  for (let i = 2; i < argsLength; i++) {
    const arg = args[i];
    if (arg[0] === '-') {
      const name = getArgName(arg);
      if (args[i + 1] === undefined) {
        parsedArgs.named[name] = true;
      } else {
        parsedArgs.named[name] = args[i + 1];
      }
      i++;
    } else {
      parsedArgs.positional.push(arg);
    }
  }

  return parsedArgs;
}

function getArgName(arg) {
  let sliceIndex;
  const isWordArg = arg[1] === '-';
  if (isWordArg) {
    if (arg.length === 3) throw new Error(`expected ${arg} to be a word arg`);
    sliceIndex = 2;
  } else {
    if (arg.length > 2) throw new Error(`expected ${arg} to be a single char arg`);
    sliceIndex = 1;
  }
  return arg.slice(sliceIndex, arg.length);
}
