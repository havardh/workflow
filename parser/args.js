// @flow
/* eslint-disable no-console */
import { InvalidArgument, InvalidConfig } from '../error';

export type Args = string | Array<string>

function validateArguments(config, args) {
  if (typeof config === 'object' && config !== null && config.length) {
    if (config.length !== args.length) {
      throw new InvalidArgument(
        `multiple arguments (${config.length}) '${config.join(' ')}' wanted,
        got (${args.length}) '${args.join(' ')}'`,
        config,
        args,
      );
    }
  } else if (typeof config === 'string') {
    if (args.length === 0) {
      throw new InvalidArgument(
        `single argument '${config}' wanted, got: '${String(args)}'`,
        config,
        args,
      );
    } else if (args.length > 1) {
      throw new InvalidArgument(
        `single argument '${config}' wanted, got: '${args.length}'`,
        config,
        args,
      );
    }
  } else {
    throw new InvalidConfig(
      `expected config.args to be of type 'string' or 'array, got: ${typeof config}'`, config);
  }
}

export default function parseArgs(
  config: Args,
  args: Array<string>,
): {[string]: string} { // eslint-disable-line no-shadow
  validateArguments(config, args);

  if (typeof config === 'object' && config.length && config.length === args.length) {
    const argsObject = {};
    config.forEach((arg) => { argsObject[String(arg)] = args.shift(); });
    return argsObject;
  } else if (typeof config === 'string' && args.length === 1) {
    return { [config]: args[0] };
  }
  return {};
}
