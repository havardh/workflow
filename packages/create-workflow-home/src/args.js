import { dev } from 'shared/env';

function usage() {
  console.log(`Initialize workflow-home directory

Usage:
  npm init workflow-home [target]
  npx create-workflow-home [target]
  yarn create workflow-home [target]

Initializes the current directory or target if given
with the create-workflow-home initializer.
`);
}

export default function parse(args) {
  if (args.length < 1) {
    console.error('Invalid number of arguments');
    console.log();
    usage();
    process.exit(1);
  }

  const command = args[0];

  if (dev && command.endsWith('node')) {
    console.log('Running dev mode');

    const [node, cmd, targetFolder] = args;

    if (args.length === 2) {
      return {};
    } else if (args.length === 3) {
      return { targetFolder };
    } else {
      console.error('Invalid arguments');
      console.log();
      usage();
      process.exit(1);
    }
  }

  switch (command) {
    case 'npm':
      if (args.length !== 3 || args.length !== 4) {
        console.error('Invalid number of arguments');
        console.log();
        usage();
        process.exit(1);
      } else if (args[1] !== 'init') {
        console.error('Unsupported npm command:', args[1]);
        console.log();
        usage();
        process.exit(1);
      } else if (args[2] !== 'workflow-home') {
        console.error('Unknown initializer name:', args[1]);
        console.log();
        usage();
        process.exit(1);
      } else if (args.length === 3) {
        return {};
      } else if (args.length === 4) {
        return { targetFolder: args[3] };
      }
      break;
    case 'yarn':
      if (args.length !== 3 || args.length !== 4) {
        console.error('Invalid number of arguments');
        console.log();
        usage();
        process.exit(1);
      } else if (args[1] !== 'create') {
        console.error('Unsupported yarn command:', args[1]);
        console.log();
        usage();
        process.exit(1);
      } else if (args[2] !== 'workflow-home') {
        console.error('Unknown initializer name:', args[1]);
        console.log();
        usage();
        process.exit(1);
      } else if (args.length === 3) {
        return {};
      } else if (args.length === 4) {
        return { targetFolder: args[3] };
      }
      break;
    case 'npx':
      if (args.length !== 2 || args.length !== 3) {
        console.error('Invalid number of arguments');
        console.log();
        usage();
        process.exit(1);
      } else if (args[2] !== 'create-workflow-home') {
        console.error('Unknown initializer name:', args[1]);
        console.log();
        usage();
        process.exit(1);
      } else if (args.length === 3) {
        return {};
      } else if (args.length === 4) {
        return { targetFolder: args[3] };
      }
  }
  console.log('Invalid arguments');
  console.log();
  usage();
  process.exit(1);
}
