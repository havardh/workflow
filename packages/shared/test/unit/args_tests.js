/* eslint-env jest */
import { args, getArgName, removeScriptName } from '../../args';

describe('args(args)', () => {
  it('should parse single character arguments correctly', () => {
    const input = ['hello', 'world', '-c', 'config', '-f', 'filename', 'otherargs'];
    const expectedOutput = {
      named: { c: 'config', f: 'filename' },
      positional: ['hello', 'world', 'otherargs'],
    };
    expect(args(input)).toEqual(expectedOutput);
  });

  it('should parse word arguments correctly', () => {
    const input = ['hello', 'world', '--config', 'config', '--filename', 'filename', 'otherargs'];
    const expectedOutput = {
      named: { config: 'config', filename: 'filename' },
      positional: ['hello', 'world', 'otherargs'],
    };
    expect(args(input)).toEqual(expectedOutput);
  });

  it('should throw an error if value is not passed for arg', () => {
    const input = ['hello', 'world', '-c'];
    expect(() => args(input)).toThrowError();
  });
});

describe('getArgName(arg)', () => {
  it('should extract the argument name for single character arg', () => {
    expect(getArgName('-c')).toBe('c');
  });

  it('should extract the argument name for single word arg', () => {
    expect(getArgName('--config')).toBe('config');
  });

  it('should extract the argument name for multi word arg', () => {
    expect(getArgName('--some-long-name')).toBe('some-long-name');
  });

  it('should throw an error if - is succeded by a word', () => {
    expect(() => getArgName('-config')).toThrowError();
  });

  it('should throw an error if -- is succeded by a single character', () => {
    expect(() => getArgName('--c')).toThrowError();
  });
});

describe('removeScriptName(args)', () => {
  it('should remove when run with node', () => {
    const input = ['/usr/bin/node', 'script.js', 'other', 'args'];
    const expectedOutput = ['other', 'args'];
    expect(removeScriptName(input)).toEqual(expectedOutput);
  });

  it('should remove when run with workflow', () => {
    const input = ['workflow', 'other', 'args'];
    const expectedOutput = ['other', 'args'];
    expect(removeScriptName(input)).toEqual(expectedOutput);
  });
});
