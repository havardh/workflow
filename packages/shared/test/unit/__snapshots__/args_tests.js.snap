// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`argument name should throw an error if - is succeded by a word 1`] = `"expected -config to be a single char arg"`;

exports[`argument name should throw an error if -- is succeded by a single character 1`] = `"expected --c to be a word arg"`;
